WIP: JDK-8217477: Implement attenuation-coefficients for PointLight

ProggerFox opened the pull request:
Adding attenuation factors to the ES2-pipeline of the JavaFX 3D-API

ProggerFox committed:
Add ES2-attenuation

Adding attenuation factors to the JavaFX 3D-API

ProggerFox committed:
Small fixes #1

Added SetRange opeator

brcolow commented:
I'm not sure if this is super WIP or not but it's pretty non-standard to add comments to all the changes with your handle "FalcoTheBold" as that's what git/hg history is for. Not a big deal just pointing it out this superficial (non-code related) detail.

ProggerFox committed:
Small fixes #2

- Added missing shader-files
- Changed default values of attenuation-factors

FalcoTheBold commented:
@brcolow They can indeed be considered super WIP, these comments are just reminders from the time when i worked on the code apart from my usual work station and are supposed to be removed/replaced later of course.

ProggerFox committed:
Fixes for AppVeyor-build

ProggerFox committed:
Delete LightBase.java

ProggerFox committed:
Delete D3DMeshView.java

ProggerFox committed:
Fixes for AppVeyor-build

ProggerFox committed:
Fixes for AppVeyor-build #2

ProggerFox committed:
[Test] Sources for Java-side

ProggerFox committed:
Small fixes #3

- added missing members for native MeshView-struct
- changed 'la' to correct data-type in native setPointLight-operation

ProggerFox committed:
IIncluded Java+D3D-patch

ProggerFox committed:
Revert "IIncluded Java+D3D-patch"

This reverts commit 3556ec77c3dbb3a4ce43a2d55fcebf37183aeff1.

ProggerFox committed:
Revert "Small fixes #3"

This reverts commit 18e58fec4ae06adcd9addad7f8f52e4122743399.

ProggerFox committed:
Revert "Revert "IIncluded Java+D3D-patch""

This reverts commit 0a5d380c510eed291aaa449f04cc39f26cfa6c3b.

ProggerFox committed:
Revert "Revert "Small fixes #3""

This reverts commit 444cc945e12f1bda4d3bfd208ec2a4de0173d90c.

ProggerFox committed:
Update D3DPhongMaterial

ProggerFox committed:
Updated D3D-files

ProggerFox committed:
Update D3DPhongMaterial.h

nlisker commented on .idea/misc.xml:
```diff
@@ -1,11 +1,7 @@
-  </component>
   <component name="IdProvider" IDEtalkID="9BC0DE4049720C1246C2D3B835C6F7C3" />
-  <component name="ProjectRootManager" version="2" languageLevel="JDK_1_8" assert-keyword="true" jdk-15="true" project-jdk-name="1.8" project-jdk-type="JavaSDK">
+  <component name="ProjectRootManager" version="2" languageLevel="JDK_1_8" default="false" project-jdk-name="1.8" project-jdk-type="JavaSDK">
```
I don't think changes to IDE files belong in this patch. Is it necessary?

FalcoTheBold replied:
No, these were set by my IDE when i imported OpenJFX it seems, but i changed these back already. Thus it will be resetted within the next commits.

nlisker commented on modules/javafx.graphics/src/main/resources/com/sun/prism/es2/glsl/main1Light.frag:
```diff
@@ -82,8 +85,12 @@ void main()
     s = pow(clamp(dot(-refl, l), 0.0, 1.0), power)*lights[0].color.rgb;
+	
+	//FalcoTheBold - added attenuation calculation for a single light
+	float dist = (lights[0].pos.xyz - gl_Position) / lights[0].atten.range;
```
I don't understand this. If the distance from the pixel to the light is `lights[0].pos.xyz - gl_Position`, why divide by the range? If that distance is larger than the range, the light should have no effect. I don't see how this happens.

FalcoTheBold replied:
Honestly, tiredness got the better of me that time and i was led too much by findings in other forums regarding how to use the range in these places. Truly an error on my part there. There will be a conditional element regarding the affected range instead in the next commit as well.

nlisker reviewed (changes requested):
OpenJFX uses spaces in stead of tabs, please change this.

Moreover, there are many spacing changes, such as in D3DContext.java and I'm not sure at all these are within the style conventions of OpenJFX. Kevin would have to comment on this I think.

The comments with your name and what you changed might be useful for reviewers, just remember to remove them later.

nlisker commented:
Did you test with the test code I linked to you that both implementations produce the same results?

kevinrushforth commented:
> Moreover, there are many spacing changes, such as in D3DContext.java and I'm not sure at all these are within the style conventions of OpenJFX. Kevin would have to comment on this I think.

I think these should be reverted. This sort of wholesale reformatting is discouraged because it makes changes harder to review (and can sometimes cause merge conflicts).

ProggerFox committed:
ES2 Attenuation-logic changes

FalcoTheBold commented:
@nlisker 

> Did you test with the test code I linked to you that both implementations produce the same results?

Not yet, but i plan to do just that shortly.
I admit i worked a little hastly in order to finish this issue before the next incoming flood of customer support will steal the remaining time from me.

Regarding

> Moreover, there are many spacing changes, such as in D3DContext.java and I'm not sure at all these are within the style conventions of OpenJFX. Kevin would have to comment on this I think.

I took the file as it is from the sources you provided and the IDE seems to have no issues with the indents (i use spaces, no tabs) or the likes.
My own additions indeed need some reworking, albeit there are indent issues at places in the sources i didn't touched.
Surely this is not a common thing with the sources in OpenJFX? (not saying i wouldn't fix them though)

ProggerFox committed:
Formatting changes

- Tabs to spaces
- line at end of sources
- removal of "Falco"-comments
- changed policy of line separator

ProggerFox committed:
Update GLContext.c

nlisker commented on modules/javafx.graphics/src/main/native-prism-d3d/D3DLight.h:
```diff
@@ -36,13 +36,16 @@ class D3DLight {
 
 #endif  /* D3DLIGHT_H */
 
+
```
Remove empty lines.

nlisker commented on modules/javafx.graphics/src/main/java/com/sun/prism/es2/ES2Context.java:
```diff
@@ -449,8 +449,11 @@ void setAmbientLight(long nativeHandle, float r, float g, float b) {
-        glContext.setPointLight(nativeHandle, index, x, y, z, r, g, b, w);
+    /*
+     * FacloTheBold: - new operator, added parameters for attenuation coefficients
+     */
```
Remove this comment if you removed the rest.

nlisker commented on modules/javafx.graphics/src/main/resources/com/sun/prism/es2/glsl/main1Light.frag:
```diff
@@ -65,26 +66,31 @@ varying vec4 lightTangentSpacePositions[3];
+        float att = 1.0 / (lights[0].atten.ca + lights[0].atten.la * dist + lights[0].atten.qa * (dist * dist)));
 
-    gl_FragColor = vec4(clamp(rez, 0.0, 1.0) , diffuse.a);
+        vec3 rez = (ambientColor+d) * (att * (diffuse.xyz + s*specular.rgb));
```
I think that `att` should multiply `d` and `s` (same as multiplying `lights[0].color.rgb`). The same is done in the d3d pipeline.

nlisker reviewed (changes requested):
Additional comments:

* `.idea/misc.xml` has been changed again.
* Undo the formatting (adding extra spaces) changes to: `MeshView.java`, `D3DContext.java`, `LightBase.java` and `D3DMeshView.java`.

nlisker commented:
> I took the file as it is from the sources you provided and the IDE seems to have no issues with the indents (i use spaces, no tabs) or the likes.
My own additions indeed need some reworking, albeit there are indent issues at places in the sources i didn't touched.
Surely this is not a common thing with the sources in OpenJFX? (not saying i wouldn't fix them though)

The sources I uploaded do not change indentation. I'm not sure what indentation issues you are referring to.

FalcoTheBold commented:
> The sources I uploaded do not change indentation. I'm not sure what indentation issues you are referring to.

FalcoTheBold commented:
You see a Github-rookie at its finest, sorry for the inconveniences.
I plan to squash the commits soon, but since @nlisker is now listed as reviewer, how should i go on about the squashing?

nlisker commented:
> May i know which IDE you used for working on your sources?

I use Eclipse.

ProggerFox committed:
Reworked shader-sources

ProggerFox committed:
Shader-sources: removed extra brackets

nlisker commented:
The implementation looks promising. Can you test with the app I uploaded that we get the same visuals?

nlisker commented on modules/javafx.graphics/src/main/resources/com/sun/prism/es2/glsl/main1Light.frag:
```diff
@@ -65,26 +66,30 @@ varying vec4 lightTangentSpacePositions[3];
 
-    vec3 rez = (ambientColor+d) * diffuse.xyz + s*specular.rgb;
-    rez += apply_selfIllum().xyz;
+        vec3 rez = (ambientColor+d) * (diffuse.xyz + s*specular.rgb);
```
I'm not sure that this change to the formula is correct. `d` should only multiply the diffuse color I think. Same for the other shaders.

nlisker commented:
@FalcoTheBold How's it going? We are slowly running out of time to get all the planned changes into 13. We need to leave enough time for reviewers too.

FalcoTheBold commented:
> @FalcoTheBold How's it going? We are slowly running out of time to get all the planned changes into 13. We need to leave enough time for reviewers too.

Hello there.
At the moment i am in quite a pinch with my regular work, finding little time to work on this and my other projects. Nonetheless i am on it to test it on the Linux-systems, just lately i had a sprint with my colleague on another work, thus taking less time on the issues here. This will change in a short while however, i will update you then.

kevinrushforth commented:
As announced in [this message](https://mail.openjdk.java.net/pipermail/openjfx-dev/2019-September/023551.html), the official `jfx` repository is moving from [hg.openjdk.java.net/openjfx/jfx-dev/rt](https://hg.openjdk.java.net/openjfx/jfx-dev/rt) to [github.com/openjdk/jfx](https://github.com/openjdk/jfx).

This [sandbox repository](https://github.com/javafxports/openjdk-jfx) is being retired on 1-Oct-2019. You will need to migrate your WIP pull request to the openjdk/jfx repo if you want to continue working on it. Here are [instructions for migrating your pull request](https://mail.openjdk.java.net/pipermail/openjfx-dev/2019-September/023560.html). The updated [CONTRIBUTING.md](https://github.com/openjdk/jfx/blob/master/CONTRIBUTING.md) doc has additional information on how to proceed.

Once you have done this, it would be helpful to add a comment with a pointer to the new PR.

NOTE: since this is a feature / enhancement request it needs to be discussed on the [openjfx-dev](mailto:openjfx-dev@openjdk.java.net) mailing list if you want it to be considered.

nlisker commented:
PR on the Skara repo: https://github.com/openjdk/jfx/pull/43
